File: frontend/src/api/avatar.ts
import http from "./http";

export interface AvatarConfig {
  id: string;
  user_id: string;
  bot_id: string | null;
  name: string;
  description: string;
  render_type: "3d" | "live2d" | "sprite";
  scene_config: Record<string, any>;
  action_mapping: Record<string, any>;
  is_default: boolean;
  avatar_assets?: AvatarAsset[];
  bot?: { id: string; name: string } | null;
  created_at: string;
  updated_at: string;
}

export interface AvatarAsset {
  id: string;
  avatar_id: string;
  asset_id: string;
  role: string;
  config: Record<string, any>;
  sort_order: number;
  asset?: {
    id: string;
    name: string;
    category: string;
    format: string;
    thumbnail_key: string;
  };
}

export interface CreateAvatarParams {
  name: string;
  description?: string;
  render_type: "3d" | "live2d" | "sprite";
  scene_config?: Record<string, any>;
  action_mapping?: Record<string, any>;
}

export interface UpdateAvatarParams {
  name?: string;
  description?: string;
  scene_config?: Record<string, any>;
  action_mapping?: Record<string, any>;
}

export interface BindAssetParams {
  asset_id: string;
  role: string;
  config?: Record<string, any>;
  sort_order?: number;
}

export const avatarApi = {
  list: (page = 1, pageSize = 20) =>
    http.get("/avatars", { params: { page, page_size: pageSize } }),
  get: (id: string) => http.get(`/avatars/${id}`),
  create: (data: CreateAvatarParams) => http.post<{ data: AvatarConfig }>("/avatars", data),
  update: (id: string, data: UpdateAvatarParams) => http.put(`/avatars/${id}`, data),
  delete: (id: string) => http.delete(`/avatars/${id}`),
  bindBot: (id: string, botId: string) =>
    http.post(`/avatars/${id}/bind-bot`, { bot_id: botId }),
  bindAsset: (id: string, data: BindAssetParams) =>
    http.post(`/avatars/${id}/bind-asset`, data),
  unbindAsset: (id: string, assetId: string) =>
    http.delete(`/avatars/${id}/assets/${assetId}`),
  updateActionMapping: (id: string, actionMapping: Record<string, any>) =>
    http.put(`/avatars/${id}/action-mapping`, { action_mapping: actionMapping }),
};

File: backend/internal/model/avatar.go
package model

import (
	"time"

	"gorm.io/gorm"
)

// AvatarConfig represents a virtual avatar configuration entity.
type AvatarConfig struct {
	ID            string         `gorm:"type:char(26);primaryKey" json:"id"`
	UserID        string         `gorm:"type:char(26);index;not null;column:user_id" json:"user_id"`
	BotID         string         `gorm:"type:char(26);index;column:bot_id" json:"bot_id"`
	Name          string         `gorm:"type:varchar(128);not null" json:"name"`
	Description   string         `gorm:"type:text" json:"description"`
	RenderType    string         `gorm:"type:varchar(16);not null;column:render_type" json:"render_type"`
	SceneConfig   string         `gorm:"type:jsonb;default:'{}';column:scene_config" json:"scene_config"`
	ActionMapping string         `gorm:"type:jsonb;default:'{}';column:action_mapping" json:"action_mapping"`
	IsDefault     bool           `gorm:"default:false;column:is_default" json:"is_default"`
	CreatedAt     time.Time      `gorm:"autoCreateTime" json:"created_at"`
	UpdatedAt     time.Time      `gorm:"autoUpdateTime" json:"updated_at"`
	DeletedAt     gorm.DeletedAt `gorm:"index" json:"deleted_at,omitempty"`

	User         User          `gorm:"foreignKey:UserID" json:"-"`
	Bot          *Bot          `gorm:"foreignKey:BotID" json:"-"`
	AvatarAssets []AvatarAsset `gorm:"foreignKey:AvatarID" json:"avatar_assets,omitempty"`
}

// TableName specifies the database table name for the AvatarConfig model.
func (AvatarConfig) TableName() string {
	return "avatar_configs"
}

// AvatarAsset represents the many-to-many relationship between avatars and assets.
type AvatarAsset struct {
	ID        string `gorm:"type:char(26);primaryKey" json:"id"`
	AvatarID  string `gorm:"type:char(26);not null;uniqueIndex:idx_avatar_asset;column:avatar_id" json:"avatar_id"`
	AssetID   string `gorm:"type:char(26);not null;uniqueIndex:idx_avatar_asset;column:asset_id" json:"asset_id"`
	Role      string `gorm:"type:varchar(32);not null" json:"role"`
	Config    string `gorm:"type:jsonb;default:'{}'" json:"config"`
	SortOrder int    `gorm:"default:0;column:sort_order" json:"sort_order"`

	Avatar AvatarConfig `gorm:"foreignKey:AvatarID;constraint:OnDelete:CASCADE" json:"-"`
	Asset  Asset        `gorm:"foreignKey:AssetID" json:"asset,omitempty"`
}

// TableName specifies the database table name for the AvatarAsset model.
func (AvatarAsset) TableName() string {
	return "avatar_assets"
}
